File: Solutions/543_Diameter_of_Binary_Tree/543_Diameter_of_Binary_Tree.go
func diameterOfBinaryTree(root *TreeNode) int {
    var max int
    deepestPath(root, &max)
    return max
}

func deepestPath(t *TreeNode, max *int) int {
    if t == nil {
		return 0
	}
	left := deepestPath(t.Left, max)
	right := deepestPath(t.Right, max)
    sum := maxValue(left, right) + 1
	*max = maxValue(*max, left + right)
	return sum
}

func maxValue(a, b int) int {
    if a > b {
        return a
    }
    return b
}

//Another
func diameterOfBinaryTree(root *TreeNode) int {
    res := 0
    visit(root, &res)
    return res
}

func visit(root *TreeNode, res *int) int {
    if root == nil {
        return 0
    }
    
    left := visit(root.Left, res)
    right := visit(root.Right, res)
    
    *res = max(*res, left + right)
    return 1 + max(left, right)
    
}

func max(a, b int) int {
    if a > b {
        return a
    }
    
    return b
}

//Fastest
func diameterOfBinaryTree(root *TreeNode) int {
	_, ret := maxDepth(root, 1)

	return ret - 1
}

func maxDepth(root *TreeNode, ret int) (int, int) {
	if root == nil {
		return 0, ret
	}

	left, lmax := maxDepth(root.Left, ret)
	right, rmax := maxDepth(root.Right, ret)

	ret = max(lmax, rmax)
	ret = max(ret, left+right+1)

	return max(left, right) + 1, ret
}

func max(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

File: Solutions/543_Diameter_of_Binary_Tree/543_Diameter_of_Binary_Tree.c
//Second fastest
#define MAX(a, b) ((a) > (b) ? (a) : (b))

int bt_diameter(struct TreeNode* root, int *diameter)
{   
    if (root == NULL)
        return 0;
    
    int left_h = bt_diameter(root->left, diameter);
    int right_h = bt_diameter(root->right, diameter);
    int curr_dia = left_h + right_h;
    
    *diameter = MAX(*diameter, curr_dia);
    return 1 + MAX(left_h, right_h);
}

int diameterOfBinaryTree(struct TreeNode* root)
{
    int diameter = 0;
    bt_diameter(root, &diameter);
    return diameter;
}

//Another
int height(struct TreeNode* root, int *ans){
    if (root == NULL) return 0; 
  
    int left = height(root->left, ans);
    int right = height(root->right, ans);
        
    *ans = fmax(*ans, left+right);
    return 1 + fmax(left, right);
}

int diameterOfBinaryTree(struct TreeNode* root){
    int ans = 0;
    height(root, &ans);
    return ans;
}

//Fastest
int find_diameter(struct TreeNode* root, int* diameter) {
    if (!root)
        return 0;
    
    int left = find_diameter(root->left, diameter);
    int right = find_diameter(root->right, diameter);
    if (*diameter < left + right)
        *diameter = left + right;
    return left > right ? left + 1 : right + 1;
}


int diameterOfBinaryTree(struct TreeNode* root){
    int diameter = 0;
    find_diameter(root, &diameter);
    return diameter;
}